### PR TITLE
Allow to delete online workspaces and factorize workspace name input

### LINKED ISSUE #613: Text input of the `Create a new workspace` modal is buggy
**Describe the bug**
Text input of the `Create a new workspace` modal is buggy

**To Reproduce**
1. Open the create a new workspace modal
2. Enter a name
3. Edit this name by replacing some of the letters
4. The cursor always goes back to the end of the text: difficult to edit a previously typed name

### PR DESCRIPTION
**Pull request by bbenoist**
# Feature

## Work performed

* Factorize workspace creation modal and workspace settings page (fixes #613)
* Don't delete workspaces in DB but use a `deletedAt` property (implements #606)
* Allow to delete a workspace from the workspace settings page (implements #606)
* Set values for workspace `createdAt` and `updatedAt` at the repository side instead of from the resolvers
* Generate slugs at the repository side instead of from the resolvers

## Results

* Cursor works OK when typing a workspace name
* You can delete a workspace
* `createdAt` and `updatedAt` for a workspace and dataset are automatically managed by their repositories. You don't have to set them at each time you call `<repository>.update`
* New workspace and dataset slugs cannot differ from the name its been generated from

## Problems encountered

* The `deletedAt` mechanism is painful to implement and maintain (details in _"Caveats"_ below).

* I couldn't move the `createdAt` and `updatedAt` management for images at the repository-side for now since the values are set when initializing templates or uploading images. We have to change this behavior so that we can guarantee that the dates will always be correct.

## Caveats

* `deletedAt`
  * We need to prevent updates to the `deletedAt` property from outside of the workspace repository.
  * You can edit a resource whose workspace has been deleted. I could have done the same checks as for workspaces everywhere in the code but, at this point, I prefer that we decide whether to keep or not the `deletedAt` mechanism.
  * We have to filter workspaces with `deletedAt: null` in each DB query on the workspace entity.
  * Deleted workspaces resources are still exposed by our servers when referring to their ID. We would have to filter entities which depends on deleted workspaces. Hence needing additional (or more complex) queries.
  * Workspaces are not deleted anymore. We need to either implement a dedicated garbage collector, or remove the use of `deletedAt`.
  * `Workspace.name` is no more a source of truth when querying for workspaces (since we change the associated slug to `${slug}-${id}` when removing the workspace

## Resolved issues

* Fixes #613
* Implements #606 

## Newly raised issues

* Implement a DB garbage collector or remove the `deletedAt` mechanism

### CODE REVIEW

File: typescript/web/src/components/settings/workspace/danger-zone/delete-workspace-warning.tsx [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        <WarningSign />
+        <Box w="full">
+          <UnorderedList>
+            <ListItem>
+              {"Deleting the "}
+              <b>{workspace?.name}</b>
+              {" workspace will delete all of its images and labels"}
+            </ListItem>
+            <ListItem>The workspace plan will be canceled</ListItem>
+            <ListItem>Every members will lose access to the workspace</ListItem>
```
Comment (etiennedupont) [inline typescript/web/src/components/settings/workspace/danger-zone/delete-workspace-warning.tsx:35]:
```suggestion
            <ListItem>Every member will lose access to the workspace</ListItem>
```

File: typescript/db/src/repository/workspace.ts [hunk 2]
```diff
@@ -47,73 +54,99 @@ export const getWorkspace: Repository["workspace"]["get"] = async (
...
+) => {
+  const db = await getPrismaClient();
+  const data = await getWorkspace(where, user);
+  if (isNil(data)) {
+    throw new Error("Could not find workspace");
+  }
+  const { stripeCustomerId, slug } = data;
+  await db.workspace.update({
+    where: castObjectNullsToUndefined(where),
+    data: { deletedAt: new Date().toISOString(), slug: `${slug}-${data.id}` },
```
Comment (etiennedupont) [inline typescript/db/src/repository/workspace.ts:147]: You also need to change the `name` of the workspace, or you'll end up having conflicting name issues (see general comment)

**Review by etiennedupont — CHANGES_REQUESTED** [review]
The feature works overall, I just noticed that you can't create a workspace when one with the same name was deleted already (see remark inside the code). 

I'm not sure to understand the need to move some logic from common-resolvers to the repository side directly? I think that our goal was initially to keep repositories as simple as possible to anticipate backend plugins in the future.

As a side note:
I think there are some important differences in philosophy and structure as compared to the rest of the codebase, which makes me think that we probably need to (decide and) enforce some guidelines. e.g:
- Usage of context vs querying vs passing props down (ex: typescript/web/src/components/settings/workspace/context.ts)
- Functional programming vs Object-oriented (ex: typescript/db/src/utils/stripe.ts)
- Files structure `x.y.ts` vs `y/x.ts` (ex: typescript/web/src/components/workspace-name-input/workspace-name-input.stories.tsx)

In particular, it feels a bit messy to me to see such different file structures cohabiting together, I think we should try to avoid that.
Screenshot 2021-12-23 at 15 46 41

File: typescript/common-resolvers/src/utils/validate-workspace-name.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  "posts",
+  "pricing",
+  "request-access",
+  "settings",
+  "static",
+  "terms",
+  "thank-you",
+  "website",
+  "workspaces",
+];
```
[the remark this answers is not in the document]
Comment (bbenoist) [reply typescript/common-resolvers/src/utils/validate-workspace-name.ts:28]: I've only moved an existing constant (which was called `forbiddenWorkspaceSlugs`) 😉 

About your remark, the problem here is that we're in the `common-resolvers` package which is not aware about the routes defined via the `pages` file stucture in the `web` package (which `common-resolvers` cannot depend on).

File: data/mutations.graphql [hunk 4]
```diff
@@ -40,6 +40,7 @@ type Mutation {
     where: WorkspaceWhereUniqueInput!
     data: WorkspaceUpdateInput!
   ): Workspace
+  deleteWorkspace(where: WorkspaceWhereUniqueInput!): Workspace
```
[the remark this answers is not in the document]
Comment (bbenoist) [reply data/mutations.graphql:43]: It's by design. I've made it so you cannot access deleted workspaces from outside of the workspace repository.
The fact that we don't actually delete workspaces is an internal implementation detail (_based on performance problem assumptions but that's another debate_). It's not really an API feature.

File: typescript/common-resolvers/src/utils/validate-workspace-name.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (etiennedupont) [reply typescript/common-resolvers/src/utils/validate-workspace-name.ts:28]: Yes I was rather thinking about doing an action in the CI, which is aware of the whole repo, but that's a bit intricate for the reward

File: data/mutations.graphql (hunk 4 above)
[the remark this answers is not in the document]
Comment (etiennedupont) [reply data/mutations.graphql:43]: Since you explicitly use this field in `db`, that causes typing errors, because the repository types are unified

**Comment by bbenoist** [comment]
Thank you for this code-review!

I'll fix the issue occurring when a workspace has already been deleted. Nice catch!

Below are my answers to the general remarks you've submitted.

Indeed, they are parts which will have to be included in wider discussions but I wanted to leave detailed-enough explanations here so that anyone can share their mind if they see something wrong.

## Placing some logic in repositories instead of resolvers

Overall, the problem is that we don't have a business layer between the resolvers and the repositories.
As an example, in a common SOLID server architecture, you have some `Service` classes which are reusable everywhere across the other services, resolvers and repositories.
These service layers are responsible of the business logic and the resolvers layers are only responsible of exposing it to the API protocol(s) we need to support. They wouldn't call any repository directly.

More concretely you would have: `WorkspaceResolver -> WorkspaceService -> WorkspaceRepository`

It's kind of long to explain all of the benefits but here are some good frameworks based on such architectures which explains how it works:
  * [Nest](https://docs.nestjs.com/providers)
  * [Loopback](https://loopback.io/doc/en/lb4/Inside-LoopBack-Application.html#api-implementation)

Since we don't have such an intermediary layer, I had to choose between placing logic in the repositories or in the resolvers.

Below is the rationale behind each choice.

#### `deletedAt`

Because I had to implement the `deletedAt` mechanism, I had a choice to make:

1. Check whether `deletedAt` is `null` everywhere we use the workspace repository (i.e. in all calls to `get`, `list`, `update`, etc.). Not even in the repository resolver but everywhere it's being used.
2. Implement the verification at the repository-side and make it so workspaces with `deletedAt` being set are considered as inexistent from outside.

Since we only use of this property because of low-level performance considerations, and because solution 1 would require a lot more code everywhere in the resolvers codebase, I've decided to go over solution 2.

#### `createdAt` / `updatedAt`

As for the `deletedAt` mechanism, these properties are low-level values whose management has to be centralized at a low-level so that we don't miss setting and updating them. In some systems, these values are even set automatically by the engine.

#### Stripe

This one is clearly debatable :sweat_smile:

At the beginning, I wanted the repository to be responsible of the `slug` management. Since we consider slugs as being like a unique identifier derived from the `name` property, I wanted to enforce that rule by calling the `getSlug` function at the `create` and `update` (only if needed) steps of the repository.

However, as Stripe accounts needs the workspace slug to be created, I had the choice to:

1. Wait for the workspace to be created by the repository and get the resulting slug. Then, create the Stripe account and send an update to the repository with the Stripe customer ID.
2. Create the Stripe account from the repository after the slug got generated and before the workspace gets created in the DB.
3. Same as 2 but with a callback as argument to the repository.

I've chosen 2 but I agree that it's not a perfect solution. Stripe shouldn't have to be used at such a low-level.

To continue with the idea of the `Service` layer, this is typically the kind of things (slug + Stripe) that I would have put in the `create`, `delete` and `update` functions of a `WorkspaceService`.

## Usage of context

When you want to split a component into multiple parts, you often have to forward some state values and some functions acting on these values.

While passing-down props is the simplest, it can be very verbose when you have a deep level of decomposition. Your code becomes complex if you want to forward only the necessary props to each part of your DOM.
One way to simplify this is to use a context such as React.Context or the one provided by Zustand.
By using a context, you only use the values and functions where necessary and don't have to forward props anymore. Overall, the code gets cleaner and simpler to understand.

#### Why did I use React Context over Zustand

While I could have used Zustand to manage the shared states, I wanted the code to be as functional and immutable as possible.

The problem with the way zustand works is that:
* It relies on a singleton. You can't clearly decide of the scope of your context.
* [Without wrapping a factory function into a React Context, it cannot have initial values](https://github.com/pmndrs/zustand#react-context) (unless you rely on `useEffect` and trigger re-renders).
* Unless you use some hacks with `useEffect`, you cannot share values coming from hooks.

As such, I've decided to go with the most pure (functionally speaking) solution which is having a custom context provider where all of the values computation is made.

Thanks to this, the whole dependency chain for calculated values remains centralized, purely functional and immutable. And it would not trigger unnecessary re-renders.

## Functional programming vs Object-oriented programming

I don't see the point of enforcing the use of one vs another. We must use the paradigm which fits the most our needs. Sometimes, it's cleaner to use OOP instead of passing an object to every function acting on it. That's kind of the same problem than with using contexts vs props propagation :wink:

For the Stripe class, I've decided to go this way because :

* We have to take care of the Stripe pointer. Managing it in a class enforces the usage of Stripe in isolated, single-responsibility functions.

* It leaves the possibility to remove the usage of a singleton without having to refactor the whole code.

* That's kind of the standard way I'm used to architecture such things. I have `Service` classes which can be injected (via dependency-injection) to other `Service`, `Repository` or `Resolver` classes.

Given that we use a singleton for now, I can still keep it "functional" and change the code to avoid the need to pass by an object. It's not that important...

## File naming

One of the things which got me loose time when working on this PR was that there was 3 files with the same name (for the component, the test and the story). Combined with the long path where the file was placed, I couldn't determine which file was the one I wanted to open when using `Cmd+P` in VS Code. I had to open one of the suggested files, see its content and reopen another one if it didn't match the first time, praying that the second time will end up with the file I initially wanted.
Meanwhile, when using the `.test` and `.story` convention, I could navigate to each file without having to lose time checking in which directory it is located.

As I've renamed the test and storybook files, the `__tests__` and `__stories__` directories became meaningless since they only contained 1 file whose name was already distinct-enough from the component file to be moved alongside it.

Because they are unit-tests and "unit-stories", I think that it makes sense to keep them alongside the component they cover. Moreover, it simplifies the directory tree structure.

About the need to have a common rule, I'd say that while I would keep the `__tests__` directory for complex tests, I think that it would be simpler and cleaner to use such a convention for the unit-tests since they should be tied to a single source file. You have a better navigation and it kind of enforces consistent file names and the single-responsibility principle.

Following the same way of thinking, the `.context` extension has been used because it represents the context of the top-level component. As for `.test` and `.story`, I find it appropriate to share the same base name than the component it manages instead of just calling it `context.tsx`.

And the `.query` and `.mutation` have been used because it would allow us to generate all of the corresponding I/O types using a `**/*.(query|mutation).ts` wildcard in the future: [_TypeScript GraphQL Code Generator – Generate GraphQL Types with Apollo Codegen Tutorial_](https://www.apollographql.com/blog/tooling/apollo-codegen/typescript-graphql-code-generator-generate-graphql-types/)

**Pull request merged**